File: java/testsuite/OfflineExamples/PolygonTriangulation/src/PolygonTriangulation.java
// Java classes
import java.io.File;
import java.util.ArrayList;
import java.util.List;

// Vitral classes
import vsdk.toolkit.environment.geometry.geometricProcessing.polygonTriangulation.MonotoneDecompositionTriangulator;
import vsdk.toolkit.environment.geometry.surface.polygon.Polygon2D;
import vsdk.toolkit.io.image.ImagePersistence;
import vsdk.toolkit.media.RGBImageUncompressed;
import vsdk.toolkit.media.RGBPixel;

// Application classes
import io.PolygonReader;
import model.PolygonModel;
import model.RenderTransform;
import options.CommandLineOptions;
import render.PolygonRasterizer;
import render.TriangleRasterizer;

public class PolygonTriangulation {
    private static CommandLineOptions
    parseCommandLineOptions(String[] args)
    {
        try {
            CommandLineOptions commandLineOptions = CommandLineOptions.parse(args);
            if ( commandLineOptions.shouldShowHelp() ) {
                CommandLineOptions.printUsage();
                return null;
            }
            return commandLineOptions;
        }
        catch ( IllegalArgumentException exception ) {
            System.err.println(exception.getMessage());
            CommandLineOptions.printUsage();
            return null;
        }
    }

    private static Polygon2D
    loadInputPolygon(CommandLineOptions commandLineOptions) throws Exception
    {
        PolygonReader polygonReader = new PolygonReader();
        return polygonReader.read(commandLineOptions.getInputFileName());
    }

    private static List<MonotoneDecompositionTriangulator.Triangle>
    triangulatePolygon(Polygon2D inputPolygon)
    {
        MonotoneDecompositionTriangulator triangulator =
            new MonotoneDecompositionTriangulator();
        List<MonotoneDecompositionTriangulator.Triangle> triangles =
            new ArrayList<>();
        triangulator.triangulate(inputPolygon, triangles);
        return triangles;
    }

    private static void
    printTriangles(List<MonotoneDecompositionTriangulator.Triangle> triangles)
    {
        for ( int triangleIndex = 0; triangleIndex < triangles.size(); triangleIndex++ ) {
            MonotoneDecompositionTriangulator.Triangle triangle =
                triangles.get(triangleIndex);
            System.out.printf("triangle #%d: %d %d %d%n", triangleIndex,
                triangle.a, triangle.b, triangle.c);
        }
    }

    private static RGBImageUncompressed
    createWorkingImage(PolygonModel model)
    {
        RGBImageUncompressed image = new RGBImageUncompressed();
        image.init(model.getImageWidth(), model.getImageHeight());
        image.createTestPattern();
        return image;
    }

    private static RGBPixel
    createBorderColor()
    {
        RGBPixel borderColor = new RGBPixel();
        borderColor.r = -1;
        borderColor.g = -1;
        borderColor.b = 0;
        return borderColor;
    }

    private static void
    renderPolygonPanel(RGBImageUncompressed image,
        PolygonModel model, RenderTransform renderTransform,
        RGBPixel borderColor)
    {
        PolygonRasterizer polygonRasterizer = new PolygonRasterizer();
        polygonRasterizer.renderSmoothFilledPolygon(image, model,
            renderTransform.getMinX(), renderTransform.getMinY(),
            renderTransform.getScale(), renderTransform.getOffsetX(),
            renderTransform.getOffsetY(), borderColor);
    }

    private static void
    renderTrianglePanel(
        RGBImageUncompressed image,
        PolygonModel model,
        List<MonotoneDecompositionTriangulator.Triangle> triangles,
        RenderTransform renderTransform, RGBPixel borderColor)
    {
        TriangleRasterizer triangleRasterizer = new TriangleRasterizer();
        triangleRasterizer.renderTriangulatedPolygon(image, model, triangles,
            renderTransform.getMinX(), renderTransform.getMinY(),
            renderTransform.getScale(),
            renderTransform.getOffsetX() + model.getZoneWidth(),
            renderTransform.getOffsetY(), borderColor);
    }

    private static void exportImage(
        RGBImageUncompressed image,
        String outputFileName) throws Exception
    {
        ImagePersistence.exportPNG(new File(outputFileName), image);
        System.out.println("Image written to: " + outputFileName);
    }

    public static void
    main(String[] args) {
        try {
            CommandLineOptions commandLineOptions = parseCommandLineOptions(args);
            if ( commandLineOptions == null ) {
                return;
            }

            Polygon2D inputPolygon = loadInputPolygon(commandLineOptions);
            PolygonModel model = commandLineOptions.toPolygonModel(inputPolygon);
            List<MonotoneDecompositionTriangulator.Triangle> triangles =
                    triangulatePolygon(inputPolygon);
            printTriangles(triangles);

            RGBImageUncompressed image = createWorkingImage(model);
            RenderTransform renderTransform =
                RenderTransform.compute(inputPolygon, model);
            RGBPixel borderColor = createBorderColor();

            renderPolygonPanel(image, model, renderTransform, borderColor);
            renderTrianglePanel(image, model, triangles, renderTransform, borderColor);
            exportImage(image, model.getOutputFileName());
        }
        catch ( Exception exception ) {
            System.err.println("PolygonTriangulation failed: " + exception.getMessage());
            exception.printStackTrace(System.err);
        }
    }
}

File: typescript/testsuite/OfflineExamples/PolygonTriangulation/src/PolygonTriangulation.ts
// Basic JDK classes
import { File, ImagePersistence } from "@vitral/fs";

// Vitral classes
import {
    IllegalArgumentException,
    MonotoneDecompositionTriangulator,
    Polygon2D,
    RGBImageUncompressed,
    RGBPixel,
} from "@vitral/base";

// Application classes
import { PolygonReader } from "./io/PolygonReader.js";
import { PolygonModel } from "./model/PolygonModel.js";
import { RenderTransform } from "./model/RenderTransform.js";
import { CommandLineOptions } from "./options/CommandLineOptions.js";
import { PolygonRasterizer } from "./render/PolygonRasterizer.js";
import { TriangleRasterizer } from "./render/TriangleRasterizer.js";

export class PolygonTriangulation {
    private static parseCommandLineOptions(args: string[]): CommandLineOptions | null {
        try {
            const commandLineOptions: CommandLineOptions = CommandLineOptions.parse(args);
            if (commandLineOptions.shouldShowHelp()) {
                CommandLineOptions.printUsage();
                return null;
            }
            return commandLineOptions;
        } catch (exception) {
            if (!(exception instanceof IllegalArgumentException)) {
                throw exception;
            }
            console.error(exception.message);
            CommandLineOptions.printUsage();
            return null;
        }
    }

    private static loadInputPolygon(commandLineOptions: CommandLineOptions): Polygon2D {
        const polygonReader: PolygonReader = new PolygonReader();
        return polygonReader.read(commandLineOptions.getInputFileName());
    }

    private static triangulatePolygon(inputPolygon: Polygon2D): MonotoneDecompositionTriangulator.Triangle[] {
        const triangulator: MonotoneDecompositionTriangulator = new MonotoneDecompositionTriangulator();
        const triangles: MonotoneDecompositionTriangulator.Triangle[] = [];
        triangulator.triangulate(inputPolygon, triangles);
        return triangles;
    }

    private static printTriangles(triangles: MonotoneDecompositionTriangulator.Triangle[]): void {
        let triangleIndex: number;
        for (triangleIndex = 0; triangleIndex < triangles.length; triangleIndex++) {
            const triangle: MonotoneDecompositionTriangulator.Triangle = triangles[triangleIndex]!;
            console.log("triangle #" + triangleIndex + ": " + triangle.a + " " + triangle.b + " " + triangle.c);
        }
    }

    private static createWorkingImage(model: PolygonModel): RGBImageUncompressed {
        const image: RGBImageUncompressed = new RGBImageUncompressed();
        image.init(model.getImageWidth(), model.getImageHeight());
        image.createTestPattern();
        return image;
    }

    private static createBorderColor(): RGBPixel {
        const borderColor: RGBPixel = new RGBPixel();
        borderColor.r = -1;
        borderColor.g = -1;
        borderColor.b = 0;
        return borderColor;
    }

    private static renderPolygonPanel(
        image: RGBImageUncompressed,
        model: PolygonModel,
        renderTransform: RenderTransform,
        borderColor: RGBPixel,
    ): void {
        const polygonRasterizer: PolygonRasterizer = new PolygonRasterizer();
        polygonRasterizer.renderSmoothFilledPolygon(
            image,
            model,
            renderTransform.getMinX(),
            renderTransform.getMinY(),
            renderTransform.getScale(),
            renderTransform.getOffsetX(),
            renderTransform.getOffsetY(),
            borderColor,
        );
    }

    private static renderTrianglePanel(
        image: RGBImageUncompressed,
        model: PolygonModel,
        triangles: MonotoneDecompositionTriangulator.Triangle[],
        renderTransform: RenderTransform,
        borderColor: RGBPixel,
    ): void {
        const triangleRasterizer: TriangleRasterizer = new TriangleRasterizer();
        triangleRasterizer.renderTriangulatedPolygon(
            image,
            model,
            triangles,
            renderTransform.getMinX(),
            renderTransform.getMinY(),
            renderTransform.getScale(),
            renderTransform.getOffsetX() + model.getZoneWidth(),
            renderTransform.getOffsetY(),
            borderColor,
        );
    }

    private static exportImage(image: RGBImageUncompressed, outputFileName: string): void {
        ImagePersistence.exportPNG(new File(outputFileName), image);
        console.log("Image written to: " + outputFileName);
    }

    public static main(args: string[]): void {
        try {
            const commandLineOptions: CommandLineOptions | null = PolygonTriangulation.parseCommandLineOptions(args);
            if (commandLineOptions === null) {
                return;
            }

            const inputPolygon: Polygon2D = PolygonTriangulation.loadInputPolygon(commandLineOptions);
            const model: PolygonModel = commandLineOptions.toPolygonModel(inputPolygon);
            const triangles: MonotoneDecompositionTriangulator.Triangle[] =
                PolygonTriangulation.triangulatePolygon(inputPolygon);
            PolygonTriangulation.printTriangles(triangles);

            const image: RGBImageUncompressed = PolygonTriangulation.createWorkingImage(model);
            const renderTransform: RenderTransform = RenderTransform.compute(inputPolygon, model);
            const borderColor: RGBPixel = PolygonTriangulation.createBorderColor();

            PolygonTriangulation.renderPolygonPanel(image, model, renderTransform, borderColor);
            PolygonTriangulation.renderTrianglePanel(image, model, triangles, renderTransform, borderColor);
            PolygonTriangulation.exportImage(image, model.getOutputFileName());
        } catch (exception) {
            const error: Error = exception as Error;
            console.error("PolygonTriangulation failed: " + error.message);
            console.error(error.stack);
        }
    }
}

PolygonTriangulation.main(process.argv.slice(2));
